blog/_posts/2019-07-14-multithreading.md

```diff
@@ -0,0 +1,464 @@
+---
+
+Software performance depends more and more on exploiting multiple processor cores.
+The [free lunch][] is still over.
+Well, we here in the Julia developer community have something of a reputation for
+caring about performance, so we've known for years that we would need a good
+story for multi-threaded, multi-core execution.
+Today we are happy to announce a major new chapter in that story.
+We are releasing an entirely new threading interface for Julia programs:
+fully general task parallelism, inspired by parallel programming systems
+like [Cilk][] and [Go][].
+
+In this paradigm, any piece of a program can be marked for execution in parallel,
+and a "task" will be started to run that code automatically on an available thread.
+A dynamic scheduler handles all the decisions and details for you.
+Here's an example of parallel code you can now write in Julia:
+
+```
+function fib(n::Int)
+    if n < 2
+        return n
+    end
+    t = @par fib(n - 2)
+    return fib(n - 1) + fetch(t)
+end
+```
+
+This, of course, is the classic highly-inefficient tree recursive implementation of
+the Fibonacci sequence --- but running on any number of processor cores!
+The line `t = @par fib(n - 2)` starts a task to compute `fib(n - 2)`, which runs in
+parallel with the following line computing `fib(n - 1)`.
+`fetch(t)` waits for task `t` to complete and gets its return value.
+
+This model of parallelism has many wonderful properties.
+I think of it as somewhat analogous to garbage collection: with GC, you
+can freely allocate objects without worrying about how it works or when and how they
+are freed.
+With task parallelism, you freely spawn tasks without worrying about where they run.
+
+The model is portable and free from low-level details.
+You don't need to explicitly start and stop threads, and you don't even need to know how
+many processors or threads there are (though you can find out if you want).
+
+The model is nestable and composable: you can start parallel tasks that call library
+functions that themselves start parallel tasks, and everything works.
+Your CPUs will not be over-subscribed with threads.
+This property is crucial for a high-level language where a lot of work is done by library
+functions.
+You need to be free to write whatever code you need --- including parallel code ---
+without worrying about how the libraries it calls are implemented.
+
+This is, in fact, the reason we are excited about this announcement: from this point on,
+multi-core parallelism is unleashed over the entire Julia package ecosystem.
+
+## Some history
+
+One of the most surprising aspects of this new feature is just how long it has been in
+the works.
+From the very beginning --- prior even to the 0.1 release --- Julia has had the `Task`
+type, providing symmetric coroutines and event-based I/O.
+So we have always had a unit of *concurrency* in the language, it just wasn't *parallel*
+(simultaneous streams of execution) yet.
```

Comment:
Does this change to "real parallelism" here have a practical effect in some computation-based IO stuff, like managing a GPU and then doing CPU operations?